### PR TITLE
feat(accessibility): snapshot the accessibility tree

### LINKED ISSUE #2033: Feature request: Inspect screen reader output
In order to help facilitate testing of a web app's accessibility, it would be great if we could inspect the screen reader output of a headless browser.

### PR DESCRIPTION
**Pull request by JoelEinbinder**
This adds `page.accessibility.snapshot()`. It serializes and returns the accessibility tree for the page. By default, uninteresting nodes are filtered out of the snapshot.

fixes #2033

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -1982,6 +1988,50 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `role` <[string]> The [role](https://www.w3.org/TR/wai-aria/#usage_intro).
+  - `name` <[string]> A human readable name for the node.
+  - `value` <[string]|[number]> The current value of the node.
+  - `description` <[string]> An additional human readable description of the node.
+  - `keyshortcuts` <[string]> Keyboard shortcuts associated with this node.
+  - `roledescription` <[string]> A human readable alternative to the role.
+  - `valuetext` <[string]> A description of the current value.
+  - `disabled` <[boolean]> Whether the node is disabled.
+  - `expanded` <[boolean]> Whether the node is expanded or collapsed.
+  - `expanded` <[boolean]> Whether the node is focused.
```
Comment (aslushnikov) [inline docs/api.md]: `focused`

File: docs/api.md [hunk 2]
```diff
@@ -1982,6 +1988,50 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
 
+### class: Accessibility
+
+The Accessibility class provides methods for inspecting Chromium's accessibility tree. The accessibility tree is used by assistive technology such as a [screen readers](https://en.wikipedia.org/wiki/Screen_reader).
+
+
+#### accessibility.snapshot([options])
+- `options` <[Object]>
+  - `interestingOnly` <[boolean]> Prune uninteresting nodes from the tree. Defaults to `true`.
+- returns: <[Promise]<[Object]>>
```
Comment (aslushnikov) [inline docs/api.md]: Returns an `AXNode` struct that has the following fields:

File: docs/api.md [hunk 3]
```diff
@@ -1982,6 +1988,50 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `checked` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `pressed` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `level` <[number]> The level of a heading.
+  - `valuemin` <[number]> The minimum value in a node.
+  - `valuemax` <[number]> the maximum value in a node.
+  - `autocomplete` <[string]> What kind of autocomplete is supported by a control.
+  - `hasPopup` <[string]> What kind of popup is currently being shown for a node.
+  - `invalid` <[string]> Whether and in what way is this node's value invalid.
+  - `orientation` <[string]> Whether the node is oriented horizontally or vertically.
+  - `children` <[Array]<[Object]>> Child nodes of this node, if any.
```
Comment (aslushnikov) [inline docs/api.md]: Child nodes, each is `AXNode`struct

File: docs/api.md [hunk 4]
```diff
@@ -1982,6 +1988,50 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
 - returns: <[string]>
 
+### class: Accessibility
+
+The Accessibility class provides methods for inspecting Chromium's accessibility tree. The accessibility tree is used by assistive technology such as a [screen readers](https://en.wikipedia.org/wiki/Screen_reader).
+
+
+#### accessibility.snapshot([options])
+- `options` <[Object]>
+  - `interestingOnly` <[boolean]> Prune uninteresting nodes from the tree. Defaults to `true`.
```
Comment (aslushnikov) [inline docs/api.md:2006]: can we elaborate on what's an "interesting"?

File: docs/api.md [hunk 5]
```diff
@@ -1982,6 +1988,50 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `hasPopup` <[string]> What kind of popup is currently being shown for a node.
+  - `invalid` <[string]> Whether and in what way is this node's value invalid.
+  - `orientation` <[string]> Whether the node is oriented horizontally or vertically.
+  - `children` <[Array]<[Object]>> Child nodes of this node, if any.
+
+Captures the current state of the accessibility tree. The returned object represents the root
+accessible node of the page.
+
+```js
+const snapshot = await page.accessibility.snapshot();
```
Comment (aslushnikov) [inline docs/api.md:2043]: is there any realistic use case? How am I supposed to use this?

File: docs/api.md [hunk 6]
```diff
@@ -3431,3 +3481,4 @@ TimeoutError is emitted whenever certain operations are terminated due to timeou
 [UnixTime]: https://en.wikipedia.org/wiki/Unix_time "Unix Time"
 [SecurityDetails]: #class-securitydetails "SecurityDetails"
 [Worker]: #class-worker "Worker"
+[Mouse]: #class-accessibility "Accessibility"
```
Comment (aslushnikov) [inline docs/api.md]: Cat.

File: lib/Accessibility.js [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+      return true;
+
+    // If it's not focusable but has a control role, then it's interesting.
+    if (this._isControl())
+      return true;
+
+    // A non focusable child of a control is not interesting
+    let parent = this._parent;
+    while (parent) {
+      if (this._isControl())
```
Comment (aslushnikov) [inline lib/Accessibility.js]: this doesn't make much sense

File: lib/Accessibility.js [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+
+class Accessibility {
+  /**
+   * @param {!Puppeteer.CDPSession} client
+   */
+  constructor(client) {
+    this._client = client;
+  }
+
+  /**
```
Comment (aslushnikov) [inline lib/Accessibility.js:63]: @param options?

File: lib/Accessibility.js [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+          break;
+        }
+        parent = parent.parent();
+      }
+    }
+    return axNodeToSerializedNode.get(root);
+  }
+}
+
+class AXNode {
```
Comment (aslushnikov) [inline lib/Accessibility.js:117]: Let's rewrite this to better outline what's going on here.
There are 2 steps:
1. Mark interesting nodes, `const interestingNodes = markInterstingNodes(tree)`. This'll be a tree traversal function
that gets a `parent: AXNode` and an `isInsideControl: boolean` parameters. This seems to be enough to compute interesting nodes.
2. Filter interesting nodes only: `const newTree = filterTree(oldTree, whitelistedNodes)`. This is just a tree filtering primitive.

For this to work, we'd need to split inhertance from AXNodes:
- Have `AXNode` that doesn't have any children/parent
- Have `TreeNode` which has just 2 fields: `axNode: AXNode` and `children: Array<TreeNode>`

File: lib/Accessibility.js [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+  constructor(client) {
+    this._client = client;
+  }
+
+  /**
+   * @return {!Promise<!SerializedAXNode>}
+   */
+  async snapshot({interestingOnly = true} = {}) {
+    const {nodes} = await this._client.send('Accessibility.getFullAXTree');
+    const root = AXNode.createTree(nodes);
```
Comment (aslushnikov) [inline lib/Accessibility.js:70]:
```js
class TreeNode {
  constructor(axnode) {
    this.axNode = axNode;
    this.children = [];
  }
}

function markInterestingNodes(node, parentNode, isInsideControl, outSet) {
  if (isInterestingNode(node.axNode, parentNode.axNode, isInsideControl))
    outSet.add(node.axNode);
  if (isLeaf(node.axNode))
    return;
  isInsideControl = isInsideControl || node.axNode.isControlNode();
  for (const child of node.children)
    markInterestingNodes(child, node, isInsideControl, outSet);
}

function filterTree(node, whitelist) {
  const children = [];
  for (const child of node.children)
    children.push(...filterTree(child, whitelist));
  if (!whitelist.has(node))
    return children;
  const newNode = new TreeNode(node.axNode);
  newNode.children = children;
  return [newNode];
}

const root = buildInitialTree();
const interesting = new Set();
markInterestingNodes(root, null, false, interesting);
const newTree = filterTree(root, interesting)[0];

```

File: docs/api.md (hunk 4 above)
Comment (JoelEinbinder) [reply docs/api.md:2006]: Done.

File: docs/api.md (hunk 2 above)
Comment (JoelEinbinder) [reply docs/api.md]: Done

File: docs/api.md (hunk 3 above)
Comment (JoelEinbinder) [reply docs/api.md]: Done.

File: docs/api.md (hunk 5 above)
Comment (JoelEinbinder) [reply docs/api.md:2043]: Added an additional example of logging the focused nodes name.

File: lib/Accessibility.js [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+      if (property.name === 'expanded')
+        this._expanded = property.value.value;
+    }
+  }
+
+  /**
+   * @return {boolean}
+   */
+  _isPlainTextField() {
+    if (this._richlyEditable)
```
Comment (aslushnikov) [inline lib/Accessibility.js:151]: can we get a test for this? a

File: lib/Accessibility.js [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+    return (role === 'LineBreak' || role === 'text' ||
+            role === 'InlineTextBox');
+  }
+
+  /**
+   * @return {boolean}
+   */
+  _hasFocusableChild() {
+    for (const child of this._children) {
+      if (child._focusable || child._hasFocusableChild())
```
Comment (aslushnikov) [inline lib/Accessibility.js]: O(n^2)!

File: test/accessibility.spec.js [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module.exports.addTests = function({testRunner, expect}) {
+  const {describe, xdescribe, fdescribe} = testRunner;
+  const {it, fit, xit} = testRunner;
+  const {beforeAll, beforeEach, afterAll, afterEach} = testRunner;
+
+  fdescribe('Accessibility', function() {
```
Comment (aslushnikov) [inline test/accessibility.spec.js]: it'll probably won't happy...

File: docs/api.md [hunk 14]
```diff
@@ -3431,3 +3502,5 @@ TimeoutError is emitted whenever certain operations are terminated due to timeou
 [UnixTime]: https://en.wikipedia.org/wiki/Unix_time "Unix Time"
 [SecurityDetails]: #class-securitydetails "SecurityDetails"
 [Worker]: #class-worker "Worker"
+[Accessibility]: #class-accessibility "Accessibility"
+[AXNode]: #accessibilitysnapshotoptions "Accessibility"
```
Comment (aslushnikov) [inline docs/api.md]: AXNode

File: docs/api.md [hunk 15]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+An example of dumping the entire accessibility tree:
+```js
+const snapshot = await page.accessibility.snapshot();
+console.log(snapshot);
+```
+
+An example of logging the focused node's name:
+```js
+const snapshot = await page.accessibility.snapshot();
+const node = findNode(snapshot, node => node.focused);
```
Comment (aslushnikov) [inline docs/api.md]: findFocusedNode?

**Review by aslushnikov — APPROVED** [review]

**Comment by JoelEinbinder** [comment]
The heuristics don't seem very accurate, especially with regards to editable nodes. I'm going to write some more tests and compare against NVDA on windows.

File: docs/api.md [hunk 16]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `expanded` <[boolean]> Whether the node is expanded or collapsed.
+  - `focused` <[boolean]> Whether the node is focused.
+  - `modal` <[boolean]> Whether the node is [modal](https://en.wikipedia.org/wiki/Modal_window).
+  - `multiline` <[boolean]> Whether the node text input supports multiline.
+  - `multiselectable` <[boolean]> Whether more than one child can be selected.
+  - `readonly` <[boolean]> Whether the node is read only.
+  - `required` <[boolean]> Whether the node is required.
+  - `selected` <[boolean]> Whether the node is selected in it's parent node.
+  - `checked` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `pressed` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
```
Comment (vsemozhetbyt) [inline docs/api.md]: The same description as above, is this intentional?

File: docs/api.md [hunk 17]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `level` <[number]> The level of a heading.
+  - `valuemin` <[number]> The minimum value in a node.
+  - `valuemax` <[number]> the maximum value in a node.
```
Comment (vsemozhetbyt) [inline docs/api.md]:
```suggestion
  - `valuemax` <[number]> The maximum value in a node.
```

File: docs/api.md [hunk 18]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `valuetext` <[string]> A description of the current value.
+  - `disabled` <[boolean]> Whether the node is disabled.
+  - `expanded` <[boolean]> Whether the node is expanded or collapsed.
+  - `focused` <[boolean]> Whether the node is focused.
+  - `modal` <[boolean]> Whether the node is [modal](https://en.wikipedia.org/wiki/Modal_window).
+  - `multiline` <[boolean]> Whether the node text input supports multiline.
+  - `multiselectable` <[boolean]> Whether more than one child can be selected.
+  - `readonly` <[boolean]> Whether the node is read only.
+  - `required` <[boolean]> Whether the node is required.
+  - `selected` <[boolean]> Whether the node is selected in it's parent node.
```
Comment (vsemozhetbyt) [inline docs/api.md]:
```suggestion
  - `selected` <[boolean]> Whether the node is selected in its parent node.
```

File: docs/api.md [hunk 19]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `checked` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `pressed` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `level` <[number]> The level of a heading.
+  - `valuemin` <[number]> The minimum value in a node.
+  - `valuemax` <[number]> the maximum value in a node.
+  - `autocomplete` <[string]> What kind of autocomplete is supported by a control.
+  - `hasPopup` <[string]> What kind of popup is currently being shown for a node.
+  - `invalid` <[string]> Whether and in what way is this node's value invalid.
```
Comment (vsemozhetbyt) [inline docs/api.md]: Not sure but maybe:
```suggestion
  - `invalid` <[string]> Whether and in what way this node's value is invalid.
```

File: docs/api.md [hunk 20]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `readonly` <[boolean]> Whether the node is read only.
+  - `required` <[boolean]> Whether the node is required.
+  - `selected` <[boolean]> Whether the node is selected in it's parent node.
+  - `checked` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `pressed` <[boolean]|[string]> Whether the checkbox is checked, or "mixed".
+  - `level` <[number]> The level of a heading.
+  - `valuemin` <[number]> The minimum value in a node.
+  - `valuemax` <[number]> the maximum value in a node.
+  - `autocomplete` <[string]> What kind of autocomplete is supported by a control.
+  - `hasPopup` <[string]> What kind of popup is currently being shown for a node.
```
Comment (vsemozhetbyt) [inline docs/api.md]: Are the option names arbitrary for the implementation? If so, maybe you can make their casing consistent? Currently, there is one camelCase here and many alllowerrcase options (which are a bit hard to read).

File: docs/api.md [hunk 21]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+  - `invalid` <[string]> Whether and in what way is this node's value invalid.
+  - `orientation` <[string]> Whether the node is oriented horizontally or vertically.
+  - `children` <[Array]<[AXNode]>> Child nodes of this node, if any.
+
+Captures the current state of the accessibility tree. The returned object represents the root accessible node of the page.
+
+> **NOTE** The Chromium accessibility tree contains nodes that go unused on most platforms and by
+most screen readers. Puppeteer will discard them as well for an easier to process tree,
+unless `interestingOnly` is set to false.
```
Comment (vsemozhetbyt) [inline docs/api.md]:
```suggestion
unless `interestingOnly` is set to `false`.
```

File: docs/api.md [hunk 22]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
 #### worker.url()
 - returns: <[string]>
 
+### class: Accessibility
```
Comment (aslushnikov) [inline docs/api.md:1991]: Accessibility is a very platform-specific thing. On different platforms, there are different screen readers that might have wildly different output.

Blink - Chrome's rendering engine - has a concept of "accessibility tree", which is than translated into different platform-specific APIs. Accessibility namespace gives users
access to the Blink Accessibility Tree.

Most of the accessibility tree gets filtered out when converting from Blink AX Tree to Platform-specific AX-Tree or by screen readers themselves. By default, Puppeteer tries to approximate this filtering, exposing only the "interesting" nodes of the tree.

File: docs/api.md [hunk 23]
```diff
@@ -1982,6 +1988,71 @@ Shortcut for [(await worker.executionContext()).evaluateHandle(pageFunction, ...
...
+```js
+const snapshot = await page.accessibility.snapshot();
+console.log(snapshot);
+```
+
+An example of logging the focused node's name:
+```js
+const snapshot = await page.accessibility.snapshot();
+const node = findNode(snapshot, node => node.focused);
+console.log(node.name);
```
Comment (vsemozhetbyt) [inline docs/api.md]: Would it throw if the `node` is `null`?

File: lib/Accessibility.js [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+      case 'scrollbar':
+      case 'slider':
+      case 'separator':
+      case 'progressbar':
+        return true;
+      default:
+        break;
+    }
+
+    if (this._role === 'combobox' && !this._expanded)
```
Comment (aslushnikov) [inline lib/Accessibility.js]: // Here and below: android heuristics

**Review by aslushnikov — APPROVED** [review]

File: docs/api.md (hunk 16 above)
Comment (JoelEinbinder) [reply docs/api.md]: Thanks.

File: docs/api.md (hunk 20 above)
Comment (JoelEinbinder) [reply docs/api.md]: Reasonable.

File: lib/Accessibility.js (hunk 12 above)
Comment (JoelEinbinder) [reply lib/Accessibility.js]: cached

**Review by aslushnikov — APPROVED** [review]

**Pull request merged**